pennylane_lightning/src/util/Util.hpp

```diff
@@ -199,87 +205,509 @@ template <class T> inline size_t dimSize(const std::vector<T> &data) {
+ * best available method.
+ *
+ * @tparam T Floating point precision type.
+ * @param mat Row-wise flatten matrix of shape m * n.
+ * @param m Number of rows of `mat`.
+ * @param n Number of columns of `mat`.
+ * @return mat transpose of shape n * m.
+ */
+template <class T>
+inline auto Transpose(const std::vector<std::complex<T>> mat, size_t m,
+                      size_t n) -> std::vector<std::complex<T>> {
+    if (mat.size() != m * n) {
+        throw std::invalid_argument("Invalid m & n for the input matrix");
+    }
+
+    std::vector<std::complex<T>> mat_t(n * m);
+    CFTranspose(mat.data(), mat_t.data(), m, n, 0, m, 0, n);
+    return mat_t;
+}
+
+/**
+ * @brief Calculates matrix-matrix product using OpenMP.
+ *
+ * @tparam T Floating point precision type.
+ * @param m_left Row-wise flatten matrix of size m * k.
+ * @param m_right Row-wise flatten matrix of size k * n.
+ * @param m_out Pre-allocated row-wise flatten matrix of size m * n.
+ * @param m Number of rows of `m_left`.
+ * @param n Number of columns of `m_right`.
+ * @param k Number of rows of `m_right`.
+ * @param transpose If `true`, requires transposed version of `m_right`.
+ *
+ * @note Consider transpose=true, to get a performance as close to CBLAS.
+ */
+template <class T>
+inline void omp_matrixMatProd(const std::complex<T> *m_left,
+                              const std::complex<T> *m_right,
+                              std::complex<T> *m_out, size_t m, size_t n,
+                              size_t k, bool transpose = false) {
+    if (!m_out) {
+        return;
+    }
+#if defined _OPENMP
+#pragma omp parallel default(none)
+#endif
+    {
+        std::size_t r, c, b;
+        if (transpose) {
+#if defined _OPENMP
+#pragma omp for
+#endif
+            for (r = 0; r < m; ++r) {
+                for (c = 0; c < n; ++c) {
+                    for (b = 0; b < k; ++b) {
+                        m_out[r * n + c] +=
+                            m_left[r * k + b] * m_right[c * n + b];
+                    }
+                }
+            }
+        } else {
+            std::size_t i, j, l;
+            std::size_t s = 2; // stride
+            std::complex<T> t;
+#if defined _OPENMP
+#pragma omp for
+#endif
+            for (r = 0; r < m; r += s) {
+                for (c = 0; c < n; c += s) {
+                    for (b = 0; b < k; b += s) {
+                        // cache-blocking:
+                        for (i = r; i < std::min(r + s, m); ++i) {
+                            for (j = c; j < std::min(c + s, n); ++j) {
+                                t = 0;
+                                for (l = b; l < std::min(b + s, k); ++l) {
+                                    t += m_left[i * k + l] * m_right[l * n + j];
+                                }
+                                m_out[i * n + j] += t;
+                            }
+                        }
+                    }
+                }
```

Comment:
This may levels of looping will annihilate the code-complexity tests. Is there another way to achieve the same logic?

Reply:
The other way around is loop-unrolling for small strides, however, this makes the function too long with many if-else conditions to consider all the corner cases.

Reply:
Ah yea, makes sense. I guess we can disable the complexity tests on these loops if/when they start to complain